lib/mayaUsd/nodes/proxyShapeBase.cpp

```diff
@@ -517,80 +570,89 @@ MayaUsdProxyShapeBase::computeInStageDataCached(MDataBlock& dataBlock)
         return MS::kSuccess;
     }
     else {
-        //
-        // Calculate from USD filepath and primPath and variantKey
-        //
-
-        // Get input attr values
-        const MString file = dataBlock.inputValue(filePathAttr, &retValue).asString();
+        // Check if we have a Stage from the Cache Id
+        const auto cacheIdNum = dataBlock.inputValue(stageCacheIdAttr, &retValue).asInt();
         CHECK_MSTATUS_AND_RETURN_IT(retValue);
+        UsdStageRefPtr usdStage;
+        const auto     cacheId = UsdStageCache::Id::FromLongInt(cacheIdNum);
+        const auto stageCached = cacheId.IsValid() && UsdUtilsStageCache::Get().Contains(cacheId);
+        if (stageCached) {
+            usdStage = UsdUtilsStageCache::Get().Find(cacheId);
```

Comment:
There may be a problem here. There are multiple UsdStageCache objects in use which could lead to incorrect behavior.

1. UsdUtilsStageCache holds a static UsdStageCache in UsdUtilsStageCache::Get().
2. UsdMayaStageCache holds two static UsdStageCache objects in its own UsdMayaStageCache::Get() implementation.

Is it possible for an input cacheId to be from a stage which is in a cache which is not the UsdUtilsStageCache? If so then I guess we fall back on the old code and try to read the stage from disk. But that might not be the correct behavior.

If we have a scene with two proxy shapes and the first proxy shape loads a file from disk (and so the stage ends up in the UsdMayaStageCache and then we connect the output cache id of that shape to the input of the other proxy shape, the 2nd proxy shape will search in the UsdUtilsStageCache and fail to find it.

I guess it works because MayaUsdProxyShapeBase::computeOutStageCacheId() inserts the stage into the UsdUtilsStageCache no matter what.

So the behavior seems fine. And the UsdStageCache doesn't duplicate the stage data, it just adds another reference to the existing stage smart pointer.

This turned into a bit of a run on train of thought with no actual action required by the code author, but I'll leave it here as a useful comment understanding how the different caches are working together safely.

Reply:
So as you deduced it won't fail. When you connect it to the output cache id it WILL put that in UsdUtilsStageCache regardless where it loads it from.

I think the MayaUSDCache is an implementation detail that might not exist moving forward for example if we use strong references for the stage.